sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala

```diff
@@ -176,44 +176,33 @@ private[sql] object JsonInferSchema {
   }
 
   /**
-   * Canonicalize inferred types, e.g., convert NullType to StringType and remove StructTypes
-   * with no fields.
+   * Recursively canonicalizes inferred types, e.g., removes StructTypes with no fields,
+   * drops NullTypes or converts them to StringType based on provided options.
    */
   private def canonicalizeType(tpe: DataType, options: JSONOptions): Option[DataType] = tpe match {
-    case at @ ArrayType(elementType, _) =>
-      val canonicalizeArrayOption = for {
-        canonicalType <- canonicalizeType(elementType, options)
-      } yield {
-        at.copy(canonicalType)
-      }
-
-      canonicalizeArrayOption.map { array =>
-        if (options.dropFieldIfAllNull && array.elementType == NullType) {
-          NullType
-        } else {
-          array
-        }
-      }
+    case at: ArrayType =>
+      canonicalizeType(at.elementType, options)
+        .map(t => at.copy(elementType = t))
 
     case StructType(fields) =>
-      val canonicalFields: Array[StructField] = for {
-        field <- fields
-        if field.name.length > 0
-        canonicalType <- canonicalizeType(field.dataType, options)
-      } yield {
-        field.copy(dataType = canonicalType)
+      val canonicalFields = fields.filter(_.name.nonEmpty).flatMap { f =>
+        canonicalizeType(f.dataType, options)
+          .map(t => f.copy(dataType = t))
       }
-
-      if (canonicalFields.length > 0) {
-        Some(StructType(canonicalFields))
-      } else if (options.dropFieldIfAllNull) {
-        Some(NullType)
+      // SPARK-8093: empty structs should be deleted
+      if (canonicalFields.isEmpty) {
+        None
       } else {
-        // per SPARK-8093: empty structs should be deleted
+        Some(StructType(canonicalFields))
+      }
+
+    case NullType =>
+      if (options.dropFieldIfAllNull) {
```

Comment:
This is the only place we check `dropFieldIfAllNull`.

Reply:
ok